src/packageurl/contrib/url2purl.py

```diff
@@ -276,3 +294,100 @@ def build_sourceforge_purl(uri):
 @purl_router.route(cargo_pattern)
 def build_cargo_purl(uri):
     return purl_from_pattern('cargo', cargo_pattern, uri)
+
+
+@purl_router.route('https?://github.com/.*')
+def build_github_purl(uri):
+    """
+    Take `uri` as input 
+    For example:
+    https://github.com/nexB/scancode-toolkit/tree/develop/plugins/scancode-ctags-macosx_10_9_intel
+    """
+    data = base_parsing(uri)
+
+    if data:
+        segments = data['segments']
+        name = data['name']
+        namespace = data['namespace']
+        version = None
+        subpath = None
+
+        #https://github.com/TG1999/fetchcode/master
+        if len(segments) >= 3  and segments[2] != 'tree':
+            version = segments[2]
+            subpath = '/'.join(segments[3:])
+
+        #https://github.com/TG1999/fetchcode/tree/master
+        if len(segments) >= 4 and segments[2] == 'tree':
+            version = segments[3]
+            subpath = '/'.join(segments[4:])
+        
+        return PackageURL(type='github', namespace=namespace, name=name, version=version, subpath=subpath)
+
+    else:
+        return
+
+ 
+@purl_router.route('https?://bitbucket.org/.*')
+def build_bitbucket_purl(uri):
+    """
+    Take `uri` as input
+    For exmaple:
+    https://bitbucket.org/TG1999/first_repo/src/master
+    """
+    data = base_parsing(uri)
+
+    if data:
+        segments = data['segments']
```

Comment:
IMHO a tuple or object might be a better returned value from your function above

Reply:
Since every URL needs different kind of parsing I think we can't return a standardize tuple, We can surely save time by not doing the base parsing every time like this.
```path = unquote_plus(urlparse(uri).path)
    segments = [seg for seg in path.split("/") if seg]

    if len(segments) <= 1:
        return

    return segments```